### ISSUE TITLE
Wrong relative offset handling in DetourBase

### ISSUE THREAD

**Issue by alemariusnexus** [DarthTon/Blackbone#418]
I believe I've found a bug in the hooking code (for what seems to be called "inline hooking"). In DetourBase::CopyOldCode(), there's a block of code that deals with instructions that have a relative offset, which will obviously become invalid if the old code is moved to somewhere else. If it finds one, it adjusts it to point to the new code location. There's a check on x64 if that new offset would be larger than 2GB, but what about instructions with smaller offsets? For example, a 'jmp short' would only work if the new offset is still within 128 bytes of the new location, which will rarely be the case. The code would just jump to a wrong address causing havoc.

There's also a problem if the relative offset is in the same code that's being "moved out of the way", because then you **don't** want to adjust it.

I don't see an easy way to fix this. I suppose it's possible to "rewrite" the instruction in question with one that supports larger offsets? Maybe there's a library that does that, because it seems like a really bad idea to roll that yourself. Until then, it's probably better to just give up and return an error code if an instruction with a relative offset is found, instead of generating wrong instructions. Or maybe it's possible to check the new offset against the instruction's offset size?

**Comment by DarthTon** [comment]
Indeed, it does not handle short jumps. Detecting them should not be a problem and there is already a fallback when we can't copy enough function bytes.
